src/reuse/lint.py

```diff
@@ -8,329 +8,260 @@
-    for i, lic in enumerate(sorted(report.deprecated_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Licenses without file extension:"))
-    for i, lic in enumerate(sorted(report.licenses_without_extension)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Missing licenses:"))
-    for i, lic in enumerate(sorted(report.missing_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Unused licenses:"))
-    for i, lic in enumerate(sorted(report.unused_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Used licenses:"))
-    for i, lic in enumerate(sorted(report.used_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Read errors: {count}").format(count=len(report.read_errors)))
-    out.write("\n")
-
-    out.write("* ")
-    out.write(
-        _("Files with copyright information: {count} / {total}").format(
-            count=file_total - len(report.files_without_copyright),
-            total=file_total,
-        )
-    )
-    out.write("\n")
 
-    out.write("* ")
-    out.write(
-        _("Files with license information: {count} / {total}").format(
-            count=file_total - len(report.files_without_licenses),
-            total=file_total,
+    def custom_serializer(obj):
+        """Custom serializer for the dictionary output of ProjectReport
+
+        :param obj: Object to be serialized
+        """
+        if isinstance(obj, PosixPath):
+            return str(obj)
+        if isinstance(obj, set):
+            return list(obj)
+        raise TypeError(
+            f"Object of type {obj.__class__.__name__} is not JSON serializable"
         )
+
+    return json.dumps(
+        report.to_dict(),
+        indent=2,
+        # Serialize sets to lists
+        default=custom_serializer,
+        sort_keys=True,
     )
-    out.write("\n")
 
 
-def add_arguments(parser):
-    """Add arguments to parser."""
-    parser.add_argument(
-        "-q", "--quiet", action="store_true", help="Prevents output"
-    )
+def lint(report: ProjectReport, formatter=format_plain, out=sys.stdout) -> bool:
+    """Lints the entire project
+
+    :param report: Dictionary holding formatted ProjectReport data
+    :param formatter: Callable that formats the data dictionary
+    :param out: Where to output
+    """
 
+    out.write(formatter(report))
 
-def run(args, project: Project, out=sys.stdout):
+    data = report.to_dict()
+    result = data["summary"]["compliant"]
```

Comment:
I'm not certain this function should exist, or at least not exist in this shape. It does the following things:

- Print lint result using a formatter.
- Return whether the lint is compliant or not, which I'm not sure is related to the printing.
  - In order to do this, `to_dict()` is called on the report, which seems like a design error to me. There should be a Python function or property `is_compliant` on ProjectReport.

If we fix the remark in the third bullet point, then this becomes effectively:

```python
out.write(formatter(report))
return report.is_compliant
```

which doesn't really merit a function on its own, I think.